File: kevoree-core/org.kevoree.framework/src/main/scala/org/kevoree/framework/KevoreeXmiHelper.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.kevoree.framework

import org.kevoree.ContainerRoot
import java.util.zip.{Deflater, Inflater}
import org.slf4j.LoggerFactory
import java.io._
import org.kevoree.loader.ContainerRootLoader
import org.kevoree.serializer.ModelSerializer
<<<<<<< HEAD
=======
import xml.PrettyPrinter
>>>>>>> commit

object KevoreeXmiHelper {

  val logger = LoggerFactory.getLogger(this.getClass)

  def save(uri: String, root: ContainerRoot) = {
    val serializer = new ModelSerializer
    val result = serializer.serialize(root)
    val pp = new PrettyPrinter(3000,1)
    val fileWrite = new FileWriter(uri)
    fileWrite.append(pp.format(serializer.serialize(root)))

  }


  def load(uri: String): ContainerRoot = {
    logger.debug("load model from => " + uri)
    val localModel = ContainerRootLoader.loadModel(new File(uri));
    localModel match {
      case Some(m) => m
      case None => println("Model not loaded!"); null
    }

  }

  def loadStream(input: InputStream): ContainerRoot = {
    val localModel = ContainerRootLoader.loadModel(input);
    localModel match {
      case Some(m) => m
      case None => println("Model not loaded!"); null
    }

  }

  def saveStream(output: OutputStream, root: ContainerRoot) : Unit = {
    val serializer = new ModelSerializer
    val result = serializer.serialize(root)
    val pr = new PrintWriter(output)
    pr.print(result)
    pr.flush()
  }

  def saveCompressedStream(output: OutputStream, root: ContainerRoot) : Unit = {
    val modelStream = new ByteArrayOutputStream()
    saveStream(modelStream, root)
    val compressor = new Deflater()
    compressor.setLevel(Deflater.BEST_COMPRESSION)
    compressor.setInput(modelStream.toByteArray)
    compressor.finish()
    val buf = new Array[Byte](1024)
    while (!compressor.finished()) {
      val count = compressor.deflate(buf)
      output.write(buf, 0, count)
    }
    output.flush()
  }

  def loadCompressedStream(input: InputStream): ContainerRoot = {
    val decompressor = new Inflater()
    val inputData: Array[Byte] = Stream.continually(input.read).takeWhile(-1 !=).map(_.toByte).toArray
    decompressor.setInput(inputData)
    val bos = new ByteArrayOutputStream(inputData.length)
    val buf = new Array[Byte](1024)
    while (!decompressor.finished()) {
      val count = decompressor.inflate(buf)
      bos.write(buf, 0, count)
    }
    val inputS = new ByteArrayInputStream(bos.toByteArray)
    loadStream(inputS)
  }


}

File: kevoree-core/org.kevoree.framework/src/main/java/org/kevoree/framework/KevoreeXmiHelper.kt
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kevoree.framework

import org.kevoree.ContainerRoot
import java.util.zip.Deflater
import java.util.zip.Inflater
import org.kevoree.loader.ModelLoader
import java.io.*
import org.kevoree.serializer.ModelSerializer
import org.kevoree.serializer.XMIModelSerializer
import org.kevoree.loader.XMIModelLoader

object KevoreeXmiHelper {

    fun save(uri: String, root: ContainerRoot) {
        //CHECK DIRECTORY CREATION
        val folderUri = if(uri.contains(File.separator)){
            uri.substring(0, uri.lastIndexOf(File.separator))
        } else {
            uri
        }
        val folder = File(folderUri)
        if (!folder.exists()){
            folder.mkdirs()
        }
        val serializer = XMIModelSerializer()
        val outputFile = File(uri);
        if(!outputFile.exists()) {
            outputFile.createNewFile()
        }
        val fop = FileOutputStream(outputFile)
        serializer.serialize(root, fop)
        fop.flush()
        fop.close()
    }

    fun saveToString(root: ContainerRoot, prettyPrint: Boolean): String {
        val serializer = XMIModelSerializer()
        val ba = ByteArrayOutputStream()
        val res = serializer.serialize(root, ba)
        ba.flush()
        val result = String(ba.toByteArray())
        ba.close()
        return result
    }

    fun loadString(model: String): ContainerRoot? {
        val loader = XMIModelLoader()
        val loadedElements = loader.loadModelFromString(model)
        if(loadedElements != null && loadedElements.size() > 0) {
            return loadedElements.get(0) as ContainerRoot;
        } else {
            return null;
        }
    }

    fun load(uri: String): ContainerRoot? {
        val loader = XMIModelLoader()
        val loadedElements = loader.loadModelFromPath(File(uri))
        if(loadedElements != null && loadedElements.size() > 0) {
            return loadedElements.get(0) as ContainerRoot;
        } else {
            return null;
        }
    }

    fun loadStream(input: InputStream): ContainerRoot? {
        val loader = XMIModelLoader()
        val loadedElements = loader.loadModelFromStream(input)
        if(loadedElements != null && loadedElements.size() > 0) {
            return loadedElements.get(0) as ContainerRoot;
        } else {
            return null;
        }

    }

    fun saveStream(output: OutputStream, root: ContainerRoot): Unit {
        val serializer = XMIModelSerializer()
        val result = serializer.serialize(root, output)
    }

    fun saveCompressedStream(output: OutputStream, root: ContainerRoot): Unit {
        val modelStream = ByteArrayOutputStream()
        saveStream(modelStream, root)
        output.write(ZipUtil.compressByteArray(modelStream.toByteArray()))
        output.flush()

    }

    fun loadCompressedStream( input: InputStream): ContainerRoot? {
        val inputS = ByteArrayInputStream(ZipUtil.uncompressByteArray(input.readBytes(input.available())))
        return loadStream(inputS)
    }


}
